### PR TITLE
Adopting according to new readability and code style

### PR DESCRIPTION
**Pull request by Maleware**
# Description

As [@NickLarsenNZ](https://github.com/NickLarsenNZ) pointed out we could get better in code style and readability. This are suggested changes of him first used in https://github.com/stackabletech/superset-operator/pull/533 and now being propagated through operators.

## Definition of Done Checklist

- Not all of these items are applicable to all PRs, the author should update this template to only leave the boxes in that are relevant
- Please make sure all these things are done and tick the boxes

```[tasklist]
# Author
- [ ] Changes are OpenShift compatible
- [ ] CRD changes approved
- [ ] CRD documentation for all fields, following the [style guide](https://docs.stackable.tech/home/nightly/contributor/docs/style-guide).
- [ ] Helm chart can be installed and deployed operator works
- [ ] Integration tests passed (for non trivial changes)
- [ ] Changes need to be "offline" compatible
```

```[tasklist]
# Reviewer
- [x] Code contains useful comments
- [ ] Code contains useful logging statements
- [x] (Integration-)Test cases ~added~ updated
- [ ] Documentation added or updated. Follows the [style guide](https://docs.stackable.tech/home/nightly/contributor/docs/style-guide).
- [ ] Changelog updated
- [ ] Cargo.toml only contains references to git tags (not specific commits or branches)
```

```[tasklist]
# Acceptance
- [ ] Feature Tracker has been updated
- [ ] Proper release label has been added
- [ ] [Roadmap](https://github.com/orgs/stackabletech/projects/25/views/1) has been updated
```

### CODE REVIEW

File: tests/templates/kuttl/overrides/31-assert.yaml [hunk 1]
```diff
@@ -4,22 +4,46 @@ kind: TestAssert
...
-  - script: |
-      POD=$(kubectl -n $NAMESPACE get pod -l app.kubernetes.io/component=restserver,app.kubernetes.io/role-group=default -o name | head -n 1 | sed -e 's#pod/##')
-      kubectl -n $NAMESPACE get pod $POD -o yaml | yq '.spec.containers[0].env[] | select (.name == "TEST_VAR_FROM_REST").value' | grep 'RESTSERVER'
-      kubectl -n $NAMESPACE get pod $POD -o yaml | yq '.spec.containers[0].env[] | select (.name == "TEST_VAR_FROM_REST_RG").value' | grep 'RESTSERVER'
-      kubectl -n $NAMESPACE get pod $POD -o yaml | yq '.spec.containers[0].env[] | select (.name == "TEST_VAR").value' | grep 'RESTSERVER_RG'
+      #!/usr/bin/env bash
+      set -euo pipefail
+
+      # POD Test Names
+      POD_MASTER=$(kubectl -n $NAMESPACE get pod -l app.kubernetes.io/component=master,app.kubernetes.io/role-group=default -o name | head -n 1 | sed -e 's#pod/##')
```
Comment (NickLarsenNZ) [inline tests/templates/kuttl/overrides/31-assert.yaml:11]:
```suggestion
      POD_MASTER=$(kubectl -n "$NAMESPACE" get pod -l app.kubernetes.io/component=master,app.kubernetes.io/role-group=default -o name | head -n 1 | sed -e 's#pod/##')
```

File: tests/templates/kuttl/overrides/31-assert.yaml [hunk 2]
```diff
@@ -4,22 +4,46 @@ kind: TestAssert
...
+      POD_REGIONSERVER_ROLE=$(kubectl -n $NAMESPACE get pod -l app.kubernetes.io/component=regionserver,app.kubernetes.io/role-group=resources-from-role -o name | head -n 1 | sed -e 's#pod/##')
+      POD_REGIONSERVER_ROLE_GROUP=$(kubectl -n $NAMESPACE get pod -l app.kubernetes.io/component=regionserver,app.kubernetes.io/role-group=resources-from-role-group -o name | head -n 1 | sed -e 's#pod/##')
+      POD_RESTSERVER=$(kubectl -n $NAMESPACE get pod -l app.kubernetes.io/component=restserver,app.kubernetes.io/role-group=default -o name | head -n 1 | sed -e 's#pod/##')
+
+      # Config Test Data
+      POD_MASTER_DATA=$(kubectl -n $NAMESPACE get pod $POD_MASTER -o yaml)
```
Comment (NickLarsenNZ) [inline tests/templates/kuttl/overrides/31-assert.yaml:17]:
```suggestion
      POD_MASTER_DATA=$(kubectl -n $NAMESPACE get pod "$POD_MASTER" -o yaml)
```

**Comment by adwk67** [comment]
TBH I find this less readable than the original.

**Review by NickLarsenNZ — CHANGES_REQUESTED** [review]
We should double quote variable expansions.

> [!TIP]
> Copy the script to a new file (eg: `file.sh`), and run `shellcheck file.sh`.

File: tests/templates/kuttl/overrides/31-assert.yaml (hunk 2 above)
Comment (Maleware) [reply tests/templates/kuttl/overrides/31-assert.yaml]: done https://github.com/stackabletech/hbase-operator/pull/553/commits/fcd0901305cb8ea33ac98cd6c53e11e902d342c2

File: tests/templates/kuttl/overrides/31-assert.yaml (hunk 1 above)
Comment (Maleware) [reply tests/templates/kuttl/overrides/31-assert.yaml]: done https://github.com/stackabletech/hbase-operator/pull/553/commits/fcd0901305cb8ea33ac98cd6c53e11e902d342c2

**Comment by NickLarsenNZ** [comment]
> TBH I find this less readable than the original.

I guess there will be no consensus, but I'll point out the problems with reviewing the original:

- Long lines wrap and make it difficult to differentiate between the pattern and variances of it (not the diff between old and new, but rather looking only at new, the vertical pattern that forms when the commands are basically the same but slightly different). Shorter lines would make this easier.
- Repeated calls to kubectl for the same thing could be combined into one. With the benefit of long kubectl command lines being reduce to a single `echo "$DESCRIPTIVE_NAME"`. But I'll admit that the abstraction could be seen as harder to read.
- Existing scripts were invalid because `kuttl` runs them via `sh -c`, without `-e`... so only the success/failure of the last command in a script was counted.
  - In these updated ones, there is a shebang, but I just realised this is meaningless inside `sh -c`... but at least we should `set -euo pipefail` (although I though `-o pipefail` was only available in bash).
- Scripts should be checked against `shellcheck`, ideally they should be files so CI checks them. Common issues I found were the lack of double quotes around variable expansion, or the use of double quotes inside double quotes which won't have the intended result.

Would it perhaps be better to write tests in a language other than a shell-script language? I think sh/bash (or at least its conventions) can make _what should be_ readable code into _difficult to read_ code.

**Comment by adwk67** [comment]
To be more specific (and hopefully a little more helpful!), I found the original change (for airflow, I think) absolutely the right thing, as we had repeated cases. But with this one I find it harder to see what I am checking for. Wrapped lines are a pain, but in these two lines 
```
POD=$(kubectl -n $NAMESPACE get pod -l app.kubernetes.io/component=master,app.kubernetes.io/role-group=default -o name | head -n 1 | sed -e 's#pod/##')
      kubectl -n $NAMESPACE get pod $POD -o yaml | yq '.spec.containers[0].env[] | select (.name == "TEST_VAR_FROM_MASTER").value' | grep 'MASTER'
```
it says to me: "identify a pod by name and rolegroup, look up the envvars and read the value for a named one". With the new version I find my eyes having to move back up the page to see what my variable-ized kubectl call was referring to.
I've no *strong* opinions about this and don't want to block the PR, and I certainly agree in changing aspects of the checks that were broken anyway (your last two points, @NickLarsenNZ).

**Comment by NickLarsenNZ** [comment]
> To be more specific (and hopefully a little more helpful!), I found the original change (for airflow, I think) absolutely the right thing, as we had repeated cases. But with this one I find it harder to see what I am checking for. Wrapped lines are a pain, but in these two lines
> 
> ```
> POD=$(kubectl -n $NAMESPACE get pod -l app.kubernetes.io/component=master,app.kubernetes.io/role-group=default -o name | head -n 1 | sed -e 's#pod/##')
>       kubectl -n $NAMESPACE get pod $POD -o yaml | yq '.spec.containers[0].env[] | select (.name == "TEST_VAR_FROM_MASTER").value' | grep 'MASTER'
> ```
> 
> it says to me: "identify a pod by name and rolegroup, look up the envvars and read the value for a named one". With the new version I find my eyes having to move back up the page to see what my variable-ized kubectl call was referring to. I've no _strong_ opinions about this and don't want to block the PR, and I certainly agree in changing aspects of the checks that were broken anyway (your last two points, @NickLarsenNZ).

Yeah good point. The jumping back up and down (or rather, the consequence of abstracting) is a pain too.

Thinking a little more about this, from the review perspective, I want to understand what the author is intending the assertions to be, then understand if that lines up with what is written. So perhaps commentary could help here too.

The wrapping is still a pain for me (and in your example above, I couldn't read the right hand side because the github UI puts a big copy button over it (doesn't happen in review, just pointing it out here).

Here is another go at spreading across multiple lines to reduce wrapping, without abstracting commands, and adding comments.

```sh
# Get the first pod name via label lookup, stripping the pod/ prefix
POD=$(
  kubectl -n "$NAMESPACE" get pod \
  -l app.kubernetes.io/component=master,app.kubernetes.io/role-group=default\
  -o name \
  | head -n 1 \
  | sed -e 's#pod/##'
)

# Assert that the pod contains 'MASTER' in the 'TEST_VAR_FROM_MASTER' env var
kubectl -n "$NAMESPACE" get pod "$POD" -o yaml \
  | yq '.spec.containers[0].env[] | select (.name == "TEST_VAR_FROM_MASTER").value'
  | grep 'MASTER'
```

The comments might seem redundant, but take the following case which might show a copy/paste error that could be picked up in review:

```sh
# Assert that the pod contains 'MASTER' in the 'TEST_VAR_FROM_MASTER' env var
kubectl -n "$NAMESPACE" get pod "$POD" -o yaml \
  | yq '.spec.containers[0].env[] | select (.name == "TEST_VAR_FROM_MASTER").value'
  | grep 'SECONDARY'
```

"oops, it looks like we are lookup up the wrong thing, or is the comment wrong?"

---

Long story short, is the example above any better for you @adwk67?

**Comment by adwk67** [comment]
That example is a big improvement, I'd say, and there may well still be cases where it makes more sense to extract specific things out into variables like your original proposal.
So I'd go with:
1. no wrapping
2. add comments
3. judicous use of variable-ized information where it aids readabilty and comprehension (subjective, I know)
(4. look at using another language - one for the back burner, probably)

**Comment by NickLarsenNZ** [comment]
> That example is a big improvement, I'd say, and there may well still be cases where it makes more sense to extract specific things out into variables like your original proposal. So I'd go with:
> 
>     1. line wrapping
> 
>     2. add comments
> 
>     3. judicous use of variable-ized information where it aids readabilty and comprehension (subjective, I know)
>        (4. look at using another language - one for the back burner, probably)

I agree with all points.

Considering I caused this whole dilemma to surface, I'll check in with @Maleware and see where I can help to achieve those points in this PR, the come back to see if that resolves the readability issue your raised above.

**Comment by NickLarsenNZ** [comment]
Small update while I'm editing things, I can definitely see how the existing style (before this PR) is ideal. It is just problematic during in the Github UI with the word wrapping during review. (_could Github please add overflow to the split-pane reviews instead of wrapping?_)

So fixing it for review makes it worse for editing, and vice-versa.

I will push up my edits anyway just so we can take a look (we can always revert them).

**Comment by NickLarsenNZ** [comment]
I have added a number of commits, each concerning a single thing. I have also added additional commit comments to give reasoning.

As I mentioned earlier, if some/all of these commits are not tolerable (please consider both writing/maintenance as well as review perspectives), I can revert them.

For what it is worth, here is the problem I had with the original change (https://github.com/stackabletech/superset-operator/pull/533) that started this set of changes. It wasn't the _changed line_ specifically, but rather the continued pattern of very long lines (which I admit are fine in my text editor) where the wrapping makes it really hard for me to review:

image

**Comment by adwk67** [comment]
> I have added a number of commits, each concerning a single thing. I have also added additional commit comments to give reasoning.

I would agree with all of these. And thank you for taking the time to give this considerable thought! (and to @Maleware for going through all the tests to make things as consistent as possible).

**Comment by NickLarsenNZ** [comment]
@Maleware shall I revert the long-opts commit (https://github.com/stackabletech/hbase-operator/pull/553/commits/fc6fd797c76689c7fd8d2fd115e42fa6e715217e)? It was more of an example showing the use of long-opts to make them clear - but they shouldn't be needed on common commands.

**Comment by Maleware** [comment]
@NickLarsenNZ Yeah I think it's morre readable without long variants.

**Review by adwk67 — APPROVED** [review]
lgtm!

File: tests/templates/kuttl/overrides/31-assert.yaml [hunk 3]
```diff
@@ -3,23 +3,108 @@ apiVersion: kuttl.dev/v1beta1
 kind: TestAssert
 timeout: 30
 commands:
+  # master, default RG
   - script: |
-      POD=$(kubectl -n $NAMESPACE get pod -l app.kubernetes.io/component=master,app.kubernetes.io/role-group=default -o name | head -n 1 | sed -e 's#pod/##')
-      kubectl -n $NAMESPACE get pod $POD -o yaml | yq '.spec.containers[0].env[] | select (.name == "TEST_VAR_FROM_MASTER").value' | grep 'MASTER'
-      kubectl -n $NAMESPACE get pod $POD -o yaml | yq '.spec.containers[0].env[] | select (.name == "TEST_VAR_FROM_MRG").value' | grep 'MASTER'
-      kubectl -n $NAMESPACE get pod $POD -o yaml | yq '.spec.containers[0].env[] | select (.name == "TEST_VAR").value' | grep 'MASTER_RG'
+      set -euo pipefail
```
Comment (adwk67) [inline tests/templates/kuttl/overrides/31-assert.yaml:8]: We might want to check this as I think something similar failed in the druid test

File: tests/templates/kuttl/overrides/31-assert.yaml [hunk 4]
```diff
@@ -3,23 +3,108 @@ apiVersion: kuttl.dev/v1beta1
...
-      kubectl -n $NAMESPACE get pod $POD -o yaml | yq '.spec.containers[0].env[] | select (.name == "TEST_VAR_FROM_REST").value' | grep 'RESTSERVER'
-      kubectl -n $NAMESPACE get pod $POD -o yaml | yq '.spec.containers[0].env[] | select (.name == "TEST_VAR_FROM_REST_RG").value' | grep 'RESTSERVER'
-      kubectl -n $NAMESPACE get pod $POD -o yaml | yq '.spec.containers[0].env[] | select (.name == "TEST_VAR").value' | grep 'RESTSERVER_RG'
+      set -euo pipefail
+
+      # Get the name of the first pod by labels
+      POD=$(
+        kubectl get pod \
+        -n "$NAMESPACE" \
+        -l app.kubernetes.io/component=restserver \
```
Comment (adwk67) [inline tests/templates/kuttl/overrides/31-assert.yaml:94]: This ORs the labels: they should be AND-ed with `-l app.kubernetes.io/component=restserver,app.kubernetes.io/role-group=default`

Comment (NickLarsenNZ) [reply tests/templates/kuttl/overrides/31-assert.yaml:94]: I had a feeling that happened, and even tested locally - but I guess my local test was flawed :facepalm:

File: tests/templates/kuttl/overrides/31-assert.yaml (hunk 3 above)
Comment (NickLarsenNZ) [reply tests/templates/kuttl/overrides/31-assert.yaml]: Fixed in 62dd3cb

File: tests/templates/kuttl/overrides/31-assert.yaml (hunk 4 above)
Comment (NickLarsenNZ) [reply tests/templates/kuttl/overrides/31-assert.yaml]: Fixed in 980a1c6

**Review by adwk67 — APPROVED** [review]
LGTM - thanks!

**Comment by Maleware** [comment]
LGTM for me too!

Thanks for your work folks :)
